Performance boost for `CircuitData` oxidation

kevinhartman opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
This PR includes:

- A new method `replace_bits` on the Rust type `CircuitData` (introduced in #10827) that takes advantage of `CircuitData`'s interned representation of `qubits` and `clbits` as vectors of indices to provide circuit remapping in constant time with respect to the number of existing instructions.
- Three new related methods, `CircuitData::map_ops`, `CircuitData::foreach_op` and `CircuitData::enumerate_ops`  that are used to invoke a Python callable for each operation in the circuit without constructing ephemeral `CircuitInstruction` instances. These methods provide a considerable performance boost for client code that was previously iterating over `CircuitData` but ignoring the `qubits` and `clbits` of yielded `CircuitInstruction`s.
- Updates to `ControlFlowBuilderBlock`'s instruction storage, which now uses `CircuitData` instead of `list`. This also includes a new method `CircuitScopeInterface.extend` which is used by `QuantumCircuit.compose` to update the currently scoped `ControlFlowBuilderBlock` all in one go.


### Details and comments

#### `CircuitData::replace_bits`
In its current form, `CircuitData` owns the `qubits` and `clbits` of a `QuantumCircuit`. As instructions are added to it, these `qubits` and `clbits` sequences are used to convert the instruction's bits from `Qubit` and `Clbit` instances to the corresponding indices, which are then stored with the `InternContext`. On access (e.g. via lookup or iteration), `CircuitData` retrieves indices from the `InternContext` and performs a lookup in the opposite direction to get back to `Qubit` and `Clbit` instances. Because we do this lookup anyway, `replace_bits` gives us a way to perform circuit remapping in constant time by simply swapping-out a `CircuitData`'s `clbits` and `qubits`.

This is used to accelerate `QuantumCircuit.compose` and `circuit_to_instruction`.

#### `CircuitData::map_ops, `CircuitData::foreach_op` and `CircuitData::enumerate_ops` 
In several cases, we don't actually need to access the `qubits` and `clbits` of `CircuitInstruction` instances while iterating over a `CircuitData`. For example, when updating a `ParameterTable`. And now with the introduction of `replace_bits`, we also have a use case for these functions to update the `operation` on conditional instructions during bit remapping.

#### `CircuitScopeInterface.extend`
This new method takes a `CircuitData` instance and uses it to extend the block's instruction listing all in one go. Because blocks now store their instructions in a `CircuitData`, a fast path has been added to `CircuitData::extend` that avoids construction of intermediate `CircuitInstruction` instances which would simply be thrown away anyway.

### Todo
- [x] Add comment with performance graphs comparing `0.45`, `oxidize-qc-data`, and this branch, `oxidize-qc-data-perf`.
- [x] Unit testing for new methods.
- [x] Minor refactoring to make `CircuitData::replace_bits` cleaner.
- [x] Update doc strings to be in Python/Sphinx form for `pyclass` methods.
- [x] Rebase and publish.

kevinhartman committed:
Optimize QC compose and circuit_to_instruction.

Use CircuitData in builder.py.

Add fastpath for CircuitData::extend(CircuitData).

Fix merge conflicts.

Add docstrings and reorder functions for an improved diff.

Optimize QuantumCircuit.copy for CircuitData.

Run formatting after merge.

Update docstrings.

Improve doc comment for replace_bits.

Fix lint.

Optimize copy.

Fix merge.

Refactor replace_bits.

Run formatter.

Use Python form docstrings in API.

Fix lint.

kevinhartman commented:
Here are some updated benchmarks for circuit building comparing memory pressure and runtime between `0.45.0`, the original `oxidize-qc-data` branch, and `oxidize-qc-data-perf` (this PR branch).

The graphs show a sweep from `depth=0` to `depth=2000`. Memory usage between `oxidize-qc-data` and `oxidize-qc-data-perf` is mostly the same, while performance is dramatically improved.

### 2Local 127Q, `flatten=False`
When `flatten` is `False`,  the optimizations to `QuantumCircuit.compose` done in this PR improve circuit building performance even beyond what we currently get in `0.45.0`.

At the upper end (`depth=2000`), `0.45` takes `~119s`, while `oxidize-qc-data-perf` takes `~99s`, a `~20s` speedup.

```py
qc = TwoLocal(
    127,
    flatten=False,
    rotation_blocks=[SXGate(), "rz", SXGate()],
    entanglement_blocks=ECRGate(),
    entanglement='pairwise',
    reps=int(sys.argv[1])
)
qc._build()
```
<img src="https://github.com/Qiskit/qiskit/assets/1374024/788b70d1-2f5a-4c0f-a321-1391146b71bb" width="400"> <img src="https://github.com/Qiskit/qiskit/assets/1374024/522940b8-4b46-495b-abda-b187661c1f8b" width="400">

### 2Local 127Q, `flatten=True`
When `flatten` is `True`, we're still slower than `0.45.0`, but we do quite a bit better than the original `oxidize-qc-data` branch.

At the upper end (`depth=2000`), `0.45` takes `~70s`, while `oxidize-qc-data-perf` takes `~79s`, a  `~9` second slowdown.


```py
qc = TwoLocal(
    127,
    flatten=True,
    rotation_blocks=[SXGate(), "rz", SXGate()],
    entanglement_blocks=ECRGate(),
    entanglement='pairwise',
    reps=int(sys.argv[1])
)
qc._build()
```
<img src="https://github.com/Qiskit/qiskit/assets/1374024/f911dcf1-adb1-4540-952e-c7332a175fee" width="400"> <img src="https://github.com/Qiskit/qiskit/assets/1374024/4cbafc46-f2de-474d-a59c-9412ce26516c" width="400">

kevinhartman commented:
To demonstrate the impact of these changes on transpilation time relative to `0.45` and `oxidize-qc-data`, here's a graph resulting from building and transpiling the following QV 127 Q circuit at depths sweeping from `0` to `1000`, with a step size of `200`.

We wouldn't expect much difference in transpilation time anyway since we convert to a `DAGCircuit` in all versions, but I thought I'd include this to validate that assumption.

```py
import sys
from qiskit import QuantumCircuit, transpile
from qiskit.circuit.library import QuantumVolume
from qiskit.providers.fake_provider import FakeWashington


backend = FakeWashington()

qv = QuantumVolume(127, int(sys.argv[1]), 1337)
qv.measure_all()

tqc = transpile(
    qv,
    backend,
    seed_transpiler=1334,
    layout_method="sabre",
    routing_method="sabre",
    optimization_level=0
)
```

![qv_transpile](https://github.com/Qiskit/qiskit/assets/1374024/de178b04-61e0-4721-b03c-38d5aaa14f30)

kevinhartman committed:
Update comment.

kevinhartman committed:
Add unit tests.

1ucian0 reviewed (commented):
The python part LGTM. But probably another pair of eyes would be better

kevinhartman committed:
Merge branch 'main' of github.com:Qiskit/qiskit into oxidize-qc-data-perf

mtreinish commented on crates/accelerate/src/quantum_circuit/circuit_data.rs:
```diff
@@ -300,6 +306,155 @@ impl CircuitData {
+    /// that appear in at least one instruction's bit lists.
+    pub fn active_bits(&self, py: Python<'_>) -> Py<PyTuple> {
+        let mut qubits: HashSet<BitType> = HashSet::new();
+        let mut clbits: HashSet<BitType> = HashSet::new();
```
Is there a `with_capacity()` we can do here to save allocations. I'd be inclined to suggest using an iterator collection here but because that would result in two iterations I think just using `with_capacity()` here instead of `new()` could save us some time. I'm thinking if we just did `with_capacity(self.qubits_native.len())` this should be fine we'd potentially over allocate if there are a lot of idle qubits, but this is just temporary as we subsequently iterate over the list for the python output and the sets will be dropped as the function exits.

```suggestion
        let mut qubits: HashSet<BitType> = HashSet::with_capacity(self.qubits_native.len());
        let mut clbits: HashSet<BitType> = HashSet::with_capacity(self.clbits_native.len());
```

kevinhartman replied:
Done in d42b4f74987f65f0c5afe09daf686ebf59f13bb9.

jakelishman commented on crates/accelerate/src/quantum_circuit/circuit_data.rs:
```diff
@@ -251,10 +251,13 @@ impl CircuitData {
+                self.qubits_native.push(bit.into_py(py));
+                self.qubits = PyList::new(py, &self.qubits_native).into_py(py);
+                Ok(())
+            })
```
Do we definitely want to return `Ok(())` if the bit is already in the circuit data?  Feels like that should be a Python-space exception - we can do that with no cost to the happy path.  (Same comment for clbits.)

Either way, we might be stretching the legibility of the chained mapping methods - the current form might be clearer as
```rust
if self.qubit_indices_native.try_insert(...).is_ok() {
    self.qubits_native.push(bit.into_py(py));
    self.qubits = ...
}
Ok(())
```

kevinhartman replied:
Improved readability in ce055ca06a69fe05189234fef77a28608c099d77.

As it is, `ControlFlowBlockBuilder.append` relies on this behavior to avoid checking if an instruction's bits are already added. If we make this an error, we'll either need to maintain parallel sets for our bit lists in the builder, or catch and swallow the would-be exception.

I don't have a strong preference.

jakelishman replied:
I think in most cases Qiskit's pretty vocal about forbidding adding bits that already exist to collections.  Do you think it would be worth making this method `strict=True` by default (raising an exception) and having the builder constructs use `strict=False`?

I'm fine with either direction: whichever you think is better for general use.

kevinhartman replied:
I like the `strict` kw-only option. Done in 62a724c7e45b9caa86f3eddf521cdea790859355.

jakelishman commented on crates/accelerate/src/quantum_circuit/circuit_data.rs:
```diff
@@ -267,27 +270,30 @@ impl CircuitData {
-            clbit_indices_native: self.clbit_indices_native.clone(),
-            qubits: PyList::new(py, &self.qubits_native).into_py(py),
-            clbits: PyList::new(py, &self.clbits_native).into_py(py),
-        })
+        let mut res = CircuitData::new(
+            py,
+            Some(self.qubits.as_ref(py)),
+            Some(self.clbits.as_ref(py)),
+            None,
+            0,
+        )?;
+        res.intern_context = self.intern_context.clone();
+        res.data = self.data.clone();
+        Ok(res)
     }
```
Unimportant, but technically there's no need for this function to return `PyResult`.

kevinhartman replied:
`CircuitData::new` sneakily returns a `PyResult` so we actually do need it in this case 😄.

jakelishman replied:
Oh sorry, I missed the ? after it.

jakelishman commented on crates/accelerate/src/quantum_circuit/circuit_data.rs:
```diff
@@ -300,6 +306,155 @@ impl CircuitData {
+            ),
+        )
+            .into_py(py)
+    }
```
This feels like we might either want to return a tuple of Python-space `set` instances, or arrange things such that the bits come out in some canonical order, to make the output deterministic.  We've had a bunch of places that have subtle bugs / non-deterministic transpilation output when we've included things like this in the past.

kevinhartman replied:
I switched to using `IndexSet` instead to get stability based on insertion order, which would be dictated by the order the bits first appear in `data`.

If we were to return a Python set, we'd still have the issue of instability I believe, since set is not ordered, correct?

Done in d42b4f74987f65f0c5afe09daf686ebf59f13bb9

jakelishman replied:
Sorry, I wasn't totally clear: I'm fine with an unordered return if we use an explicitly unordered collection as the return value (Python's `set`, for example).  Where I had a problem was that we were returning an implicitly ordered collection (`list`) that didn't actually have an order.  That's the kind of thing that's led to bugs for us before - it's very easy to think that `list` will always be ordered, but if you're working with a `set`, you know for sure that it's not.

I think `set` is probably the better return type, just thinking about how it's likely to be used; there will be cases where we'll want to iterate through all active qubits, but also commonly, I think people will want to get the set of active qubits, then query that collection for "is this object active?".  Both of those can be satisfied efficiently if we return `set`, but if we return `list`, then the second use-case ended up requiring two intermediate data structures that are immediately thrown away (the inner Rust set and the Python list).  You know better than me about how this is being used, though.

kevinhartman replied:
Ah I see. For the only current use-case, it's ideal to have these returned in a deterministic order, but I agree with you that in principle this function ought to return sets.

Changed in 375dd598eb2829bf993eae8eb9a41f7814ab861d.

jakelishman commented on crates/accelerate/src/quantum_circuit/circuit_data.rs:
```diff
@@ -300,6 +306,155 @@ impl CircuitData {
+    /// :attr:`~.CircuitInstruction.clbits` of existing instructions are
+    /// reinterpreted using the new bit sequences on access.
+    /// As such, the primary use-case for this method is to remap a circuit to
+    /// a different set of bits in O(1) time.
```
Well, it's technically `O(n_qubits + n_clbits)`, but I think you're referring to it being constant relative to the number of instructions?

kevinhartman replied:
Addressed in 21202694ecd9614587d14cb2fd1e7aba24ae672a.

jakelishman commented on qiskit/circuit/controlflow/builder.py:
```diff
@@ -251,16 +250,24 @@ def __init__(
-        self.instructions: List[CircuitInstruction] = []
-        self.qubits = set(qubits)
-        self.clbits = set(clbits)
+        self.instructions: CircuitData = CircuitData(qubits, clbits)
```
Very unimportant: the type hint is no longer necessary here, because mypy can infer it from the constructor, whereas `[]` is a subtype of all `List[T]`, so it needs a specifier.

kevinhartman replied:
Addressed in 1e4a46d1f5e17af2dc808a0df771466a7e5c2e0b.

jakelishman commented on qiskit/circuit/controlflow/builder.py:
```diff
@@ -251,16 +250,24 @@ def __init__(
+    @property
+    def clbits(self):
+        """The set of clbits associated with this scope."""
+        return set(self.instructions.clbits)
```
I'd preferentially make these methods now, not properties - they do non-trivial work on access and return mutable values.  This stuff isn't public interface, so no worries about changing it.

kevinhartman replied:
Done in df137b03865120ac905f418427aaf2fd53a5d467.

jakelishman commented on qiskit/circuit/controlflow/builder.py:
```diff
@@ -420,9 +445,14 @@ def build(
         )
 
+        # TODO: this can likely be optimized with a CircuitData.foreach_op followed
+        #   by a CircuitData.replace_bits.
```
Is it worth trying to do this immediately?

kevinhartman replied:
Done in e1d0cebfa7fbed7c6b24861e938f00b48b5e0bd3.

jakelishman commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -904,6 +904,8 @@ def compose(
             )
 
         dest = self if inplace else self.copy()
+        dest.duration = None
+        dest.unit = "dt"
```
This is probably a bugfix in principle, right - looks like `compose` wasn't doing this in all necessary situations before.  Can we add a test to enforce it in all the relevant cases, and potentially a bugfix release note?

kevinhartman replied:
I added it here because this used to happen as a result of calls to `QuantumCircuit._append`, and we now use `CircuitScopeInterface.extend`.

In theory, perhaps updating these _should_ belong in `_OuterCircuitScopeInterface.extend`, since the original behavior would be to set these only if the scope is the root `QuantumCircuit`. Thoughts?

Are there any cases you see where these should have been set before in `compose` but weren't?

jakelishman replied:
tbh, I don't really know - I don't know how these attributes are meant to work properly.

jakelishman commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -970,34 +988,40 @@ def compose(
+
+        def map_vars(op):
+            # TODO: do we really need to copy if we're not updating vars?
+            n_op = op.copy()
```
For the most part, yes - what if some other method after this were to mutate it?  As we move more operations to be singletons, the cost of this goes down hugely anyway.

jakelishman commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -2046,10 +2070,10 @@ def copy(self, name: str | None = None) -> "QuantumCircuit":
-            id(instruction.operation): instruction.operation.copy() for instruction in self._data
-        }
+        operation_copies = {}
+        self._data.foreach_op(lambda op: operation_copies.update({id(op): op.copy()}))
```
```suggestion
        def memo_copy(op):
            if (out := operation_copies.get(id(op))) is not None:
                return out
            return (operation_copies[id(op)] := op.copy())
        self._data.replace_ops(memo_copy)
```
or something like that, maybe? Better to avoid the temporary allocations of loads of one-item dicts, and this does the replacement in the same loop as the memoisation.

kevinhartman replied:
Done in b68a6aa0bb7dd3eb2453baeaf6db6a1cd5612dc5

jakelishman commented on qiskit/converters/circuit_to_instruction.py:
```diff
@@ -89,6 +89,7 @@ def circuit_to_instruction(circuit, parameter_map=None, equivalence_library=None
         equivalence_library.add_equivalence(out_instruction, target)
 
     regs = []
+    q, c = None, None
```
If we're modifying this, can we switch it to `qubits` and `clbits` - it'd be good to trim down the one-letter variables that could have many meanings (qubit / qubits / quantum register / etc?)

kevinhartman replied:
Done in f036f802cf0d06f5bb3ae990fb9ddc3786247bf6.

jakelishman reviewed (commented):
Generally this looks sensible to me.  If we're bikeshedding names, I'd perhaps suggest `map_ops` instead of `replace_ops`, just to match more with the standard Python and Rust terminology for that operation, but it's not very important.

Could you show the performance characteristics of the `Pauli.to_instruction` and `QuantumCircuit.copy` asv benchmarks between the three reference points as well, just to check everything?  The changeset here looks strongly like you've directly targetted those benchmarks, so I'm assuming you already have that data somewhere?

kevinhartman committed:
Make active_bits stable and preallocate.

kevinhartman committed:
Improve readability of add_{qu,cl}bit methods.

kevinhartman committed:
Clarify replace_bits docstring.

kevinhartman committed:
Remove unnecessary type hint.

kevinhartman committed:
Make ControlFlowBuilderBlock qubits and clbits methods.

kevinhartman committed:
Remove TODO.

kevinhartman committed:
Use single replace_ops in QuantumCircuit.copy.

kevinhartman committed:
Use memoization in circuit_to_instruction. Style fixes.

kevinhartman committed:
Merge branch 'main' of github.com:Qiskit/qiskit into oxidize-qc-data-perf

kevinhartman committed:
Merge branch 'main' of github.com:Qiskit/qiskit into oxidize-qc-data-perf

kevinhartman committed:
Add CircuitScopeInterface.extend.

This simplifies caller logic when appending an
instruction listing to a scope by leaving it up
to the scope implementation to determine what is
required (e.g. updating the parameter table only
for QuantumCircuit scopes vs just tracking bits
and instructions for a control flow block.

kevinhartman committed:
Simplify logic for front compose.

kevinhartman committed:
Add enumerate_ops, optimize block builder.

kevinhartman committed:
Rename replace_ops to map_ops.

kevinhartman commented:
@jakelishman:
> Could you show the performance characteristics of the `Pauli.to_instruction` and `QuantumCircuit.copy` asv benchmarks between the three reference points as well, just to check everything?

## `main(@181208)` vs PR
This one _should_ be mostly better, given that the original PR #10827 is already in `main` and this PR optimizes that work. We get this speedup to `main` by merging this PR (which we absolutely should do).

```
> asv compare -s 18120853a3a4f1cab461119e9643eec6029ea568 oxidize-qc-data-perf

Benchmarks that have improved:

| Change   | Before [18120853] <oxidize-qc-data-perf~4^2>   | After [c7ecf358] <oxidize-qc-data-perf>   |   Ratio | Benchmark (Parameter)                                                       |
|----------|------------------------------------------------|-------------------------------------------|---------|-----------------------------------------------------------------------------|
| -        | 217±4μs                                        | 40.1±0.3μs                                |    0.18 | circuit_construction.CircuitConstructionBench.time_circuit_copy(1, 128)     |
| -        | 199±2ms                                        | 23.5±0.3ms                                |    0.12 | circuit_construction.CircuitConstructionBench.time_circuit_copy(1, 131072)  |
| -        | 3.23±0.05ms                                    | 382±30μs                                  |    0.12 | circuit_construction.CircuitConstructionBench.time_circuit_copy(1, 2048)    |
| -        | 49.3±0.1ms                                     | 5.89±0.05ms                               |    0.12 | circuit_construction.CircuitConstructionBench.time_circuit_copy(1, 32768)   |
| -        | 33.9±1μs                                       | 26.4±2μs                                  |    0.78 | circuit_construction.CircuitConstructionBench.time_circuit_copy(1, 8)       |
| -        | 12.8±0.2ms                                     | 1.51±0.04ms                               |    0.12 | circuit_construction.CircuitConstructionBench.time_circuit_copy(1, 8192)    |
| -        | 244±2μs                                        | 52.4±1μs                                  |    0.21 | circuit_construction.CircuitConstructionBench.time_circuit_copy(14, 128)    |
| -        | 210±9ms                                        | 24.0±0.3ms                                |    0.11 | circuit_construction.CircuitConstructionBench.time_circuit_copy(14, 131072) |
| -        | 4.87±0.1ms                                     | 390±3μs                                   |    0.08 | circuit_construction.CircuitConstructionBench.time_circuit_copy(14, 2048)   |
| -        | 52.8±0.6ms                                     | 5.96±0.04ms                               |    0.11 | circuit_construction.CircuitConstructionBench.time_circuit_copy(14, 32768)  |
| -        | 73.7±0.7μs                                     | 30.7±2μs                                  |    0.42 | circuit_construction.CircuitConstructionBench.time_circuit_copy(14, 8)      |
| -        | 13.2±0.3ms                                     | 1.50±0.02ms                               |    0.11 | circuit_construction.CircuitConstructionBench.time_circuit_copy(14, 8192)   |
| -        | 225±8μs                                        | 43.5±3μs                                  |    0.19 | circuit_construction.CircuitConstructionBench.time_circuit_copy(2, 128)     |
| -        | 210±5ms                                        | 25.0±0.6ms                                |    0.12 | circuit_construction.CircuitConstructionBench.time_circuit_copy(2, 131072)  |
| -        | 3.27±0.02ms                                    | 384±5μs                                   |    0.12 | circuit_construction.CircuitConstructionBench.time_circuit_copy(2, 2048)    |
| -        | 51.3±0.6ms                                     | 6.14±0.5ms                                |    0.12 | circuit_construction.CircuitConstructionBench.time_circuit_copy(2, 32768)   |
| -        | 36.1±0.8μs                                     | 19.0±0.6μs                                |    0.53 | circuit_construction.CircuitConstructionBench.time_circuit_copy(2, 8)       |
| -        | 12.8±0.2ms                                     | 1.48±0.02ms                               |    0.12 | circuit_construction.CircuitConstructionBench.time_circuit_copy(2, 8192)    |
| -        | 299±20μs                                       | 60.3±0.6μs                                |    0.2  | circuit_construction.CircuitConstructionBench.time_circuit_copy(20, 128)    |
| -        | 208±4ms                                        | 24.4±3ms                                  |    0.12 | circuit_construction.CircuitConstructionBench.time_circuit_copy(20, 131072) |
| -        | 3.32±0.08ms                                    | 399±5μs                                   |    0.12 | circuit_construction.CircuitConstructionBench.time_circuit_copy(20, 2048)   |
| -        | 53.2±1ms                                       | 5.97±0.1ms                                |    0.11 | circuit_construction.CircuitConstructionBench.time_circuit_copy(20, 32768)  |
| -        | 96.0±2μs                                       | 39.7±1μs                                  |    0.41 | circuit_construction.CircuitConstructionBench.time_circuit_copy(20, 8)      |
| -        | 13.0±0.2ms                                     | 1.50±0.02ms                               |    0.12 | circuit_construction.CircuitConstructionBench.time_circuit_copy(20, 8192)   |
| -        | 242±4μs                                        | 45.9±1μs                                  |    0.19 | circuit_construction.CircuitConstructionBench.time_circuit_copy(5, 128)     |
| -        | 211±7ms                                        | 25.4±0.7ms                                |    0.12 | circuit_construction.CircuitConstructionBench.time_circuit_copy(5, 131072)  |
| -        | 3.22±0.06ms                                    | 403±10μs                                  |    0.13 | circuit_construction.CircuitConstructionBench.time_circuit_copy(5, 2048)    |
| -        | 51.4±0.9ms                                     | 6.11±0.05ms                               |    0.12 | circuit_construction.CircuitConstructionBench.time_circuit_copy(5, 32768)   |
| -        | 38.7±0.8μs                                     | 23.4±3μs                                  |    0.6  | circuit_construction.CircuitConstructionBench.time_circuit_copy(5, 8)       |
| -        | 12.8±0.2ms                                     | 1.54±0.02ms                               |    0.12 | circuit_construction.CircuitConstructionBench.time_circuit_copy(5, 8192)    |
| -        | 249±3μs                                        | 58.6±10μs                                 |    0.24 | circuit_construction.CircuitConstructionBench.time_circuit_copy(8, 128)     |
| -        | 213±4ms                                        | 24.4±0.1ms                                |    0.11 | circuit_construction.CircuitConstructionBench.time_circuit_copy(8, 131072)  |
| -        | 3.25±0.05ms                                    | 393±4μs                                   |    0.12 | circuit_construction.CircuitConstructionBench.time_circuit_copy(8, 2048)    |
| -        | 55.8±2ms                                       | 6.06±0.1ms                                |    0.11 | circuit_construction.CircuitConstructionBench.time_circuit_copy(8, 32768)   |
| -        | 51.6±1μs                                       | 24.8±0.3μs                                |    0.48 | circuit_construction.CircuitConstructionBench.time_circuit_copy(8, 8)       |
| -        | 13.0±0.1ms                                     | 1.51±0.04ms                               |    0.12 | circuit_construction.CircuitConstructionBench.time_circuit_copy(8, 8192)    |
| -        | 169±4μs                                        | 116±2μs                                   |    0.69 | converters.ConverterBenchmarks.time_circuit_to_instruction(1, 8)            |
| -        | 276±50μs                                       | 138±3μs                                   |    0.5  | converters.ConverterBenchmarks.time_circuit_to_instruction(2, 8)            |

Benchmarks that have stayed the same:

| Change   | Before [18120853] <oxidize-qc-data-perf~4^2>   | After [c7ecf358] <oxidize-qc-data-perf>   | Ratio   | Benchmark (Parameter)                                                |
|----------|------------------------------------------------|-------------------------------------------|---------|----------------------------------------------------------------------|
|          | 1.44±0.7ms                                     | 1.58±0.02ms                               | 1.09    | quantum_info.PauliBench.time_to_instruction(100)                     |
|          | 6.41±2ms                                       | 7.44±3ms                                  | ~1.16   | quantum_info.PauliBench.time_to_instruction(300)                     |
|          | 10.2±0.1ms                                     | 5.73±6ms                                  | ~0.56   | quantum_info.PauliBench.time_to_instruction(400)                     |
|          | 15.0±8ms                                       | 8.97±9ms                                  | ~0.60   | quantum_info.PauliBench.time_to_instruction(500)                     |

Benchmarks that have got worse:

| Change   | Before [18120853] <oxidize-qc-data-perf~4^2>   | After [c7ecf358] <oxidize-qc-data-perf>   |   Ratio | Benchmark (Parameter)                            |
|----------|------------------------------------------------|-------------------------------------------|---------|--------------------------------------------------|
| +        | 3.54±0.06ms                                    | 3.92±1ms                                  |    1.11 | quantum_info.PauliBench.time_to_instruction(200) |

```

## `stable/0.45 (@908e55)` vs PR
This compares the pre-oxidization (just Python list) `QuantumCircuit.data` to oxidized version with optimizations in place. It seems to indicate that circuit copying is faster than it was before the memory optimizations, but that `Pauli.to_instruction` is often slower.

```
> asv compare -s upstream/stable/0.45 oxidize-qc-data-perf

Benchmarks that have improved:

| Change   | Before [908e555d]    | After [c7ecf358] <oxidize-qc-data-perf>   |   Ratio | Benchmark (Parameter)                                                       |
|----------|----------------------|-------------------------------------------|---------|-----------------------------------------------------------------------------|
| -        | 108±0.7μs            | 40.1±0.3μs                                |    0.37 | circuit_construction.CircuitConstructionBench.time_circuit_copy(1, 128)     |
| -        | 94.7±0.9ms           | 23.5±0.3ms                                |    0.25 | circuit_construction.CircuitConstructionBench.time_circuit_copy(1, 131072)  |
| -        | 1.49±0.02ms          | 382±30μs                                  |    0.26 | circuit_construction.CircuitConstructionBench.time_circuit_copy(1, 2048)    |
| -        | 23.6±2ms             | 5.89±0.05ms                               |    0.25 | circuit_construction.CircuitConstructionBench.time_circuit_copy(1, 32768)   |
| -        | 5.80±0.1ms           | 1.51±0.04ms                               |    0.26 | circuit_construction.CircuitConstructionBench.time_circuit_copy(1, 8192)    |
| -        | 118±4μs              | 52.4±1μs                                  |    0.45 | circuit_construction.CircuitConstructionBench.time_circuit_copy(14, 128)    |
| -        | 96.8±0.7ms           | 24.0±0.3ms                                |    0.25 | circuit_construction.CircuitConstructionBench.time_circuit_copy(14, 131072) |
| -        | 1.49±0.02ms          | 390±3μs                                   |    0.26 | circuit_construction.CircuitConstructionBench.time_circuit_copy(14, 2048)   |
| -        | 24.1±0.7ms           | 5.96±0.04ms                               |    0.25 | circuit_construction.CircuitConstructionBench.time_circuit_copy(14, 32768)  |
| -        | 5.84±0.07ms          | 1.50±0.02ms                               |    0.26 | circuit_construction.CircuitConstructionBench.time_circuit_copy(14, 8192)   |
| -        | 119±10μs             | 43.5±3μs                                  |    0.36 | circuit_construction.CircuitConstructionBench.time_circuit_copy(2, 128)     |
| -        | 99.1±2ms             | 25.0±0.6ms                                |    0.25 | circuit_construction.CircuitConstructionBench.time_circuit_copy(2, 131072)  |
| -        | 1.65±0.1ms           | 384±5μs                                   |    0.23 | circuit_construction.CircuitConstructionBench.time_circuit_copy(2, 2048)    |
| -        | 24.6±0.3ms           | 6.14±0.5ms                                |    0.25 | circuit_construction.CircuitConstructionBench.time_circuit_copy(2, 32768)   |
| -        | 5.92±0.03ms          | 1.48±0.02ms                               |    0.25 | circuit_construction.CircuitConstructionBench.time_circuit_copy(2, 8192)    |
| -        | 127±1μs              | 60.3±0.6μs                                |    0.48 | circuit_construction.CircuitConstructionBench.time_circuit_copy(20, 128)    |
| -        | 98.3±2ms             | 24.4±3ms                                  |    0.25 | circuit_construction.CircuitConstructionBench.time_circuit_copy(20, 131072) |
| -        | 1.50±0.02ms          | 399±5μs                                   |    0.27 | circuit_construction.CircuitConstructionBench.time_circuit_copy(20, 2048)   |
| -        | 24.4±0.5ms           | 5.97±0.1ms                                |    0.24 | circuit_construction.CircuitConstructionBench.time_circuit_copy(20, 32768)  |
| -        | 5.87±0.5ms           | 1.50±0.02ms                               |    0.26 | circuit_construction.CircuitConstructionBench.time_circuit_copy(20, 8192)   |
| -        | 126±10μs             | 45.9±1μs                                  |    0.36 | circuit_construction.CircuitConstructionBench.time_circuit_copy(5, 128)     |
| -        | 98.1±0.9ms           | 25.4±0.7ms                                |    0.26 | circuit_construction.CircuitConstructionBench.time_circuit_copy(5, 131072)  |
| -        | 1.50±0.02ms          | 403±10μs                                  |    0.27 | circuit_construction.CircuitConstructionBench.time_circuit_copy(5, 2048)    |
| -        | 23.9±0.3ms           | 6.11±0.05ms                               |    0.26 | circuit_construction.CircuitConstructionBench.time_circuit_copy(5, 32768)   |
| -        | 5.90±0.03ms          | 1.54±0.02ms                               |    0.26 | circuit_construction.CircuitConstructionBench.time_circuit_copy(5, 8192)    |
| -        | 115±2μs              | 58.6±10μs                                 |    0.51 | circuit_construction.CircuitConstructionBench.time_circuit_copy(8, 128)     |
| -        | 99.4±2ms             | 24.4±0.1ms                                |    0.25 | circuit_construction.CircuitConstructionBench.time_circuit_copy(8, 131072)  |
| -        | 1.48±0.03ms          | 393±4μs                                   |    0.27 | circuit_construction.CircuitConstructionBench.time_circuit_copy(8, 2048)    |
| -        | 24.1±0.4ms           | 6.06±0.1ms                                |    0.25 | circuit_construction.CircuitConstructionBench.time_circuit_copy(8, 32768)   |
| -        | 5.87±0.1ms           | 1.51±0.04ms                               |    0.26 | circuit_construction.CircuitConstructionBench.time_circuit_copy(8, 8192)    |
| -        | 135±8μs              | 116±2μs                                   |    0.86 | converters.ConverterBenchmarks.time_circuit_to_instruction(1, 8)            |
| -        | 175±3μs              | 138±3μs                                   |    0.79 | converters.ConverterBenchmarks.time_circuit_to_instruction(2, 8)            |

Benchmarks that have stayed the same:

| Change   | Before [908e555d]    | After [c7ecf358] <oxidize-qc-data-perf>   | Ratio   | Benchmark (Parameter)                                                  |
|----------|----------------------|-------------------------------------------|---------|------------------------------------------------------------------------|
|          | 33.4±0.3μs           | 30.7±2μs                                  | 0.92    | circuit_construction.CircuitConstructionBench.time_circuit_copy(14, 8) |
|          | 19.8±0.5μs           | 19.0±0.6μs                                | 0.96    | circuit_construction.CircuitConstructionBench.time_circuit_copy(2, 8)  |
|          | 42.0±0.4μs           | 39.7±1μs                                  | 0.94    | circuit_construction.CircuitConstructionBench.time_circuit_copy(20, 8) |
|          | 23.8±0.3μs           | 24.8±0.3μs                                | 1.04    | circuit_construction.CircuitConstructionBench.time_circuit_copy(8, 8)  |
|          | 724±300μs            | 1.58±0.02ms                               | ~2.18   | quantum_info.PauliBench.time_to_instruction(100)                       |
|          | 1.79±0.03ms          | 7.44±3ms                                  | ~4.15   | quantum_info.PauliBench.time_to_instruction(300)                       |
|          | 2.25±1ms             | 5.73±6ms                                  | ~2.54   | quantum_info.PauliBench.time_to_instruction(400)                       |
|          | 2.84±0.1ms           | 8.97±9ms                                  | ~3.16   | quantum_info.PauliBench.time_to_instruction(500)                       |

Benchmarks that have got worse:

| Change   | Before [908e555d]    | After [c7ecf358] <oxidize-qc-data-perf>   |   Ratio | Benchmark (Parameter)                                                 |
|----------|----------------------|-------------------------------------------|---------|-----------------------------------------------------------------------|
| +        | 19.4±0.5μs           | 26.4±2μs                                  |    1.36 | circuit_construction.CircuitConstructionBench.time_circuit_copy(1, 8) |
| +        | 19.0±0.4μs           | 23.4±3μs                                  |    1.23 | circuit_construction.CircuitConstructionBench.time_circuit_copy(5, 8) |
| +        | 1.34±0.05ms          | 3.92±1ms                                  |    2.94 | quantum_info.PauliBench.time_to_instruction(200)                      |
```

kevinhartman committed:
Fix benchmark random_circuit.

kevinhartman commented:
I've also run benchmarks for `circuit_to_instruction` between `stable/0.45` (908e555d) and the PR branch, which show that this method is faster and scales better than before any of the `QuantumCircuit.data` oxidation work 😄:

```
Benchmarks that have improved:

| Change   | Before [908e555d]    | After [a2f2fbd7] <oxidize-qc-data-perf>   |   Ratio | Benchmark (Parameter)                                                |
|----------|----------------------|-------------------------------------------|---------|----------------------------------------------------------------------|
| -        | 132±4μs              | 120±1μs                                   |    0.91 | converters.ConverterBenchmarks.time_circuit_to_instruction(1, 8)     |
| -        | 174±20μs             | 141±2μs                                   |    0.81 | converters.ConverterBenchmarks.time_circuit_to_instruction(2, 8)     |
| -        | 270±4μs              | 201±3μs                                   |    0.75 | converters.ConverterBenchmarks.time_circuit_to_instruction(5, 8)     |
| -        | 357±6μs              | 252±5μs                                   |    0.71 | converters.ConverterBenchmarks.time_circuit_to_instruction(8, 8)     |
| -        | 1.77±0.04ms          | 1.22±0.06ms                               |    0.69 | converters.ConverterBenchmarks.time_circuit_to_instruction(53, 8)    |
| -        | 538±6μs              | 362±30μs                                  |    0.67 | converters.ConverterBenchmarks.time_circuit_to_instruction(14, 8)    |
| -        | 721±8μs              | 482±6μs                                   |    0.67 | converters.ConverterBenchmarks.time_circuit_to_instruction(20, 8)    |
| -        | 1.13±0.03ms          | 697±9μs                                   |    0.62 | converters.ConverterBenchmarks.time_circuit_to_instruction(32, 8)    |
| -        | 720±7μs              | 415±90μs                                  |    0.58 | converters.ConverterBenchmarks.time_circuit_to_instruction(1, 128)   |
| -        | 1.04±0.02ms          | 563±20μs                                  |    0.54 | converters.ConverterBenchmarks.time_circuit_to_instruction(2, 128)   |
| -        | 14.9±0.4ms           | 7.38±0.1ms                                |    0.5  | converters.ConverterBenchmarks.time_circuit_to_instruction(2, 2048)  |
| -        | 9.14±0.1ms           | 4.41±0.2ms                                |    0.48 | converters.ConverterBenchmarks.time_circuit_to_instruction(1, 2048)  |
| -        | 61.0±1ms             | 29.2±0.3ms                                |    0.48 | converters.ConverterBenchmarks.time_circuit_to_instruction(2, 8192)  |
| -        | 36.1±0.5ms           | 16.5±0.3ms                                |    0.46 | converters.ConverterBenchmarks.time_circuit_to_instruction(1, 8192)  |
| -        | 2.00±0.03ms          | 916±20μs                                  |    0.46 | converters.ConverterBenchmarks.time_circuit_to_instruction(5, 128)   |
| -        | 2.99±0.04ms          | 1.27±0.06ms                               |    0.43 | converters.ConverterBenchmarks.time_circuit_to_instruction(8, 128)   |
| -        | 10.6±0.5ms           | 4.41±0.1ms                                |    0.42 | converters.ConverterBenchmarks.time_circuit_to_instruction(32, 128)  |
| -        | 4.95±0.06ms          | 2.02±0.02ms                               |    0.41 | converters.ConverterBenchmarks.time_circuit_to_instruction(14, 128)  |
| -        | 28.5±0.5ms           | 11.6±0.2ms                                |    0.41 | converters.ConverterBenchmarks.time_circuit_to_instruction(5, 2048)  |
| -        | 77.1±2ms             | 30.7±0.4ms                                |    0.4  | converters.ConverterBenchmarks.time_circuit_to_instruction(14, 2048) |
| -        | 6.92±0.05ms          | 2.80±0.1ms                                |    0.4  | converters.ConverterBenchmarks.time_circuit_to_instruction(20, 128)  |
| -        | 126±3ms              | 50.5±1ms                                  |    0.4  | converters.ConverterBenchmarks.time_circuit_to_instruction(5, 8192)  |
| -        | 17.3±0.5ms           | 6.87±0.05ms                               |    0.4  | converters.ConverterBenchmarks.time_circuit_to_instruction(53, 128)  |
| -        | 44.5±2ms             | 17.7±0.9ms                                |    0.4  | converters.ConverterBenchmarks.time_circuit_to_instruction(8, 2048)  |

Benchmarks that have stayed the same:

| Change   | Before [908e555d]    | After [a2f2fbd7] <oxidize-qc-data-perf>   | Ratio   | Benchmark (Parameter)                                                |
|----------|----------------------|-------------------------------------------|---------|----------------------------------------------------------------------|
|          | 2.87±0.09ms          | 17.4±9ms                                  | ~6.05   | quantum_info.PauliBench.time_to_instruction(500)                     |
|          | 2.31±0.9ms           | 12.3±5ms                                  | ~5.35   | quantum_info.PauliBench.time_to_instruction(400)                     |
|          | 1.79±0.02ms          | 7.62±3ms                                  | ~4.25   | quantum_info.PauliBench.time_to_instruction(300)                     |
|          | 1.25±0.5ms           | 3.92±2ms                                  | ~3.14   | quantum_info.PauliBench.time_to_instruction(200)                     |
|          | 186±6ms              | 72.2±1ms                                  | ~0.39   | converters.ConverterBenchmarks.time_circuit_to_instruction(8, 8192)  |
|          | 741±300μs            | 43.2±800μs                                | ~0.06   | quantum_info.PauliBench.time_to_instruction(100)                     |
|          | n/a                  | n/a                                       | n/a     | converters.ConverterBenchmarks.time_circuit_to_instruction(14, 8192) |
|          | n/a                  | n/a                                       | n/a     | converters.ConverterBenchmarks.time_circuit_to_instruction(20, 2048) |
|          | n/a                  | n/a                                       | n/a     | converters.ConverterBenchmarks.time_circuit_to_instruction(20, 8192) |
|          | n/a                  | n/a                                       | n/a     | converters.ConverterBenchmarks.time_circuit_to_instruction(32, 2048) |
|          | n/a                  | n/a                                       | n/a     | converters.ConverterBenchmarks.time_circuit_to_instruction(32, 8192) |
|          | n/a                  | n/a                                       | n/a     | converters.ConverterBenchmarks.time_circuit_to_instruction(53, 2048) |
|          | n/a                  | n/a                                       | n/a     | converters.ConverterBenchmarks.time_circuit_to_instruction(53, 8192) |
```

jakelishman commented on crates/accelerate/src/quantum_circuit/circuit_data.rs:
```diff
@@ -350,6 +358,20 @@ impl CircuitData {
+    ///     func (Callable[[int, :class:`~.Operation`], None]):
+    ///         The callable to invoke.
+    #[pyo3(signature = (func))]
+    pub fn enumerate_ops(&self, py: Python<'_>, func: &PyAny) -> PyResult<()> {
```
Do you think there's any potential confusion here with the built-in name `enumerate`, which produces an iterator rather than doing the iteration?  I've seen this method called `mapi` in other languages, though I don't love that.

kevinhartman replied:
There very well may be. This function is actually unlike `map_ops` in that it does not replace the operation with the result of the supplied function. It's more like `foreach_op` but with an index.

How about either `foreach_op_i` or `foreach_op_index`?

jakelishman replied:
Oh, then definitely confusion, because I misread it as a `map`-like operation sorry!  Sounds like a good idea to include `foreach_op` in there somewhere: `foreach_op_indexed` sounds fine to me.

kevinhartman replied:
Done in 02421b6b708e807ff84d6940588d7b298599b501.

jakelishman commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -5936,6 +5931,14 @@ def append(self, instruction):
+
+        self.circuit._data.extend(data)
+        data.foreach_op(update_param)
+
```
This feels a bit prone to getting out-of-sync with `QuantumCircuit._append` here (it's already not identical).  It feels like the main performance reason for this `extend` method to exist is more efficient space reservation in `CircuitData`; perhaps it would be best to just reserve `len(data)` spaces, then call `Scope.append` in a loop to ensure everything stays in sync?

kevinhartman replied:
It's a bit more than the preallocation in this case. Under the hood, `CircuitData::extend` also has special casing for when the source data is a `CircuitData` instance, in which case it skips the creation and destruction of ephemeral `CircuitInstruction` instances during the copy. In my testing, that seemed to help with performance, though I don't have an exact comparison handy.

To be symmetrical with the relationship between `QuantumCircuit._append` and `_OuterCircuitScopeInterface.append`, we could introduce a new `QuantumCircuit._extend` method and call it from `_OuterCircuitScopeInterface.extend`. I didn't opt for that initially since `QuantumCircuit._append` is sort of part of the public API (I think), whereas a `QuantumCircuit._extend` certainly wouldn't be since it operates on the private `CircuitData` type.

jakelishman replied:
I'd definitely not want to make a split `extend`/`_extend` - that would (if anything) exacerbate the problem of having the append logic in several places.  My worry is that this is a second place where any tracking the inner `_append` primitive does needs to be synced with, whereas previously it was just one location, and this change is already introducing an inconsistency - `_append` sets `self.duration = None` and `self.unit = "dt"` on all appends, while this doesn't.  I don't know exactly _why_ `_append` does that, but it worries me that we're introducing an inconsistency (well, more replacing `compose`'s blatant disregard for the data model with something that appears the same but isn't entirely identical yet).

In a compiled language we'd obviously factor out the checks into an inlined function, but in Python we'll have a little bit of overhead.  I just tried it with a helper function in a `qiskit.qasm2.loads`-based workflow, which typically has a large dependence on the speed of `_append`, and it wasn't noticeable, though, so maybe doing this works:

```python
class QuantumCircuit:
    def _append(self, instruction):
        # ... whatever setup there is ...
        self._data.append(instruction)
        self._track_operation(instruction.operation)
        return whatever_it_is

    def _track_operation(self, operation: Operation):
        """Sync all non-data-list internal data structures for a newly tracked operation."""
        if isinstance(operation, Instruction):
            self._update_parameter_table(operation)
        self.duration = None
        self.unit = "dt"

class OuterCircuitScope(ScopeInterface):
    def extend(self, data):
        self.circuit._data.extend(data)
        data.foreach_op(self.circuit._track_operation)
```

kevinhartman replied:
That seems like a great idea to me. Done in f27d2ca01f00db8b0c2dcc49d58a7c961aaa1f5c.

jakelishman reviewed (commented):
I think this is very close to being ready to land - the performance improvements certainly look great!

Aside from two fairly minor comments here, I guess there's a little bit of merge conflict that'll need fixing, and there's a couple of not-yet-resolved comment chains I replied to in the previous review:
- https://github.com/Qiskit/qiskit/pull/11214#discussion_r1410749155
- https://github.com/Qiskit/qiskit/pull/11214#discussion_r1410761412

kevinhartman committed:
Add default strict mode for bits.

kevinhartman committed:
Return tuple of sets from active_bits.

kevinhartman committed:
Rename 'enumerate_ops' to 'foreach_op_indexed'.

kevinhartman committed:
Add _track_operation method.

kevinhartman committed:
Merge branch 'main' of github.com:Qiskit/qiskit into oxidize-qc-data-perf

kevinhartman committed:
Fix merge for global phase.

jakelishman reviewed (approved):
Thanks for all the changes on this!  All the comments I had are resolved, so I think we should merge this and start getting tested in general usage (at least on the main branch!).
